src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -389,17 +388,18 @@ void Complete()
         /// The reason for this restriction is to make this use of Reflection trim-compatible,
         /// ensuring that trimming doesn't change the application's behavior.
         /// </remarks>
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
-            Justification = "Task<T>.Result is preserved by the ILLinker because _taskGetResultMethodInfo was initialized with it.")]
         private static MethodInfo? GetTaskResultMethodInfo(Type taskType)
         {
-            MethodInfo? result = taskType.GetMethod(TaskGetResultName);
-            if (result != null && result.HasSameMetadataDefinitionAs(_taskGetResultMethodInfo))
+            try
             {
-                return result;
+                return (MethodInfo)taskType.GetMemberWithSameMetadataDefinitionAs(s_taskGetResultMethodInfo);
+            }
+            catch (ArgumentException)
```

Comment:
Would it be better to keep the original code? Is throwing and catching exception here going to be a performance problem?

Reply:
The caught exception should be rare - cases where you are returning a Task that isn't `t.GetType() == typeof(Task)` and isn't `Task<T>` or a derived type.

I'm not sure if just doing the `try` is going to be performance problem. This code is about interop'ing with JavaScript, which itself is fairly expensive.

I can revert and use the original code. But it is a pity that this was the first scenario discovered and the one written up in the API issue, and we decide that we can't use this API for this scenario.

Other options I've thought about:

1. Trying to check if the type is `Task<T>` or a derived type before calling `GetMemberWithSameMetadataDefinitionAs`, but I'm not sure how to do this efficiently and without walking the base types manually.
2. Reverting back to `GetMemberWithSameMetadataDefinitionAs` returning `null` if it isn't found. We wouldn't need to do anything special here then.

I think probably the best option in the short term is to just leave this code as it is in main. We can always change it later.

Thoughts?

Reply:
It seems like we keep dancing around the scenario of "I have a Task (or something that derives from it), and I want to get its result, and I'm ok with boxing." This seems like it should be fairly straightforward: put a first-class API on `Task` or `TaskExtensions` to provide this information. For `Task`, it could return _null_. Involving reflection for this scenario seems awfully heavyweight.

Reply:
What is Task or TaskExtensions going to do differently? It will have to use reflection to do the job.

Reply:
> without walking the base types manually.

Walking the base types manually would be like 5 lines and pretty efficient.

I am not sure whether the exceptional case is as rare as you may think - we have multiple types inheriting from Task in the framework itself. If it is really rare here, the try/catch should be ok.

It is interesting that the caller use `Array.Empty<object>()` for results that are task exactly, but this logic returns null for results that derive from Task. It suggests that the results that derive from Task may be broken. Have you actually seen a test case that hits the type deriving from Task, but not from `Task<T>` case?

Reply:
This comes up infrequently and has never risen to the level where we've wanted to do anything about it.  But in such discussions previously we've discussed having `Task<T>` implement an interface with an object result property. That would make every `Task<T>`  instantiation more expensive, though.

Reply:
> Have you actually seen a test case that hits the type deriving from Task, but not from `Task<T>` case?

I haven't in this code. But with the failing WASM tests, it looks like there are cases where types are deriving from `Task<T>`, which is causing this:

```
[22:22:48] info: browser: Console websocket connected.
[22:22:48] info: Arguments: --run,WasmTestRunner.dll,System.ComponentModel.TypeConverter.Tests.dll,-notrait,category=IgnoreForCI,-notrait,category=OuterLoop,-notrait,category=failing
[22:22:49] info: Initializing.....
[22:23:00] info: Discovering: System.ComponentModel.TypeConverter.Tests.dll (method display = ClassAndMethod, method display options = None)
[22:23:00] info: Discovered:  System.ComponentModel.TypeConverter.Tests.dll (found 1361 of 1372 test cases)
[22:23:00] info: Starting:    System.ComponentModel.TypeConverter.Tests.dll
[22:23:07] info: {"method":"console.error","payload":{},"arguments":[{}]}
[22:23:07] fail: TypeError: Cannot read property 'toString' of undefined
                     at test_exit (http://127.0.0.1:44433/runtime.js:125:41)
[22:23:07] info: Finished:    System.ComponentModel.TypeConverter.Tests.dll
[22:23:07] info: 
[22:23:07] info: === TEST EXECUTION SUMMARY ===
[22:23:07] info: Total: 7729, Errors: 0, Failed: 0, Skipped: 6, Time: 12.27358s
[22:23:07] info: 
[22:23:07] info: WASM EXIT undefined
[22:23:07] fail: Application has finished with exit code TESTS_FAILED but 0 was expected
[22:23:07] info: WASM EXIT 1
XHarness exit code: 71 (GENERAL_FAILURE)
```

AFAICT, this failure:

```
[22:23:07] fail: TypeError: Cannot read property 'toString' of undefined
                     at test_exit (http://127.0.0.1:44433/runtime.js:125:41)
````

is because the "Main" entry point to the test runner is returning a Task object that derives from `Task<T>`, and my changes to the JS Runtime code was causing the returned `Result` to be `null`.

Reply:
> a Task object that derives from Task<T>

Every `async Task` and `async Task<T>` method returns an object that derives from `Task<T>`.

`Task.ContinueWith`, `Task.Delay`, `ValueTask.AsTask()`, and more will return other types derived from `Task`.

Creating custom types that derive from `Task` and `Task<T>` is relatively rare outside of corelib, but it's pretty common inside of corelib (and those types end up getting exposed).

Reply:
> What is Task or TaskExtensions going to do differently? It will have to use reflection to do the job. (Or it will have to be non-pay-for-play.)

Reflection shouldn't be needed for this. You can pull this off with an internal interface _ITaskWithResult_ and make `Task<TResult>` implement it. You're not increasing the size of every `Task` instance, but you would be adding one entry to the vtable of each `Task<TResult>`.

```cs
public class Task<TResult> : /* existing interfaces, */ ITaskWithResult
{
    object? ITaskWithResult.GetResult() => m_result;
}

public static Task<object> TaskToTaskOfObject(this Task task)
{
    // prototype, could probably be optimized
    TaskCompletionSource<object> tcs = new();
    task.ContinueWith(static task =>
    {
        if (task.IsFaulted) { /* copy task exceptions to tcs and mark terminal */ }
        else {
            Debug.Assert(task.IsSuccessful);
            if (task is ITaskWithResult taskWithResult) { tcs.SetResult(taskWithResult.GetResult()); }
            else { tcs.SetResult(null); }
        }
    }, tcs);
    return tcs.Task;
}
```

Reply:
Or should I be reading into Steve's comment that we're averse to adding even a single virtual method to `Task<T>` given how many different types it might explode onto?

Reply:
That's the non-pay-for-play concern. If we were to decide we're fine with the cost, then the approach we've previously discussed is similar to what you outline, e.g.
```C#
public async Task<object?> AsObjectTask()
{
    await this.ConfigureAwait(false);
    return this is ITaskWithResult t ? t.ObjectResult : null;
}
```
so implementation is obviously far from hard.  Not sure exactly what you'd want to call it, or whether it'd be better hidden away somewhere as a static (non-extension) method.  Yes, this has been requested now and again, but it's not all that common a need.

Reply:
Fair enough. I saw 13 matches for `typeof(Task<>)` in source.dot.net and believed this to be a more common pattern. (To be fair, I didn't trace back many of those calls.) If this isn't as common as I thought, no worries. :)

Reply:
For my edification, how do you search for `typeof(Task<>)` in source.dot.net?

Reply:
https://source.dot.net/System.Private.CoreLib/R/5ca7b77f3df89fc6.html

Because the matches in the left pane show the reference in context (the entire source line is there), you can CTRL-F in your browser and search "typeof(Task\<\>)". Kinda hokey, but if it works, it works.

Reply:
Thanks.  Quickly skimmed the 16 results I see:
- Most of them are checking to see whether a type is a `Task` or `Task<>` in order to decide whether the method should be considered async / is task returning.
- Another large chunk are various forms of code generators trying to determine what return type to use for the generated method (e.g. CodeDOM, F#'s conversion from tasks to its async model)
- Some are looking to understand the type of the `T` in the `Task<T>` as part of API exploration / doc generation.
- A few are trying to match up the `T` to see which of a few known types it might map to, e.g. for razor page handler execution
- 1, from Microsoft.JSInterop, was about getting the object result from an arbitrary `Task<T>`